syntax = "proto3";

package hw2;
import "google/protobuf/timestamp.proto";

service MessengerServer {
  rpc Login (Request) returns (Reply) {}
  rpc Join (Request) returns (Reply) {}
  rpc Leave (Request) returns (Reply) {}
  rpc List (Request) returns (ListReply) {} 
  rpc Chat (stream Message) returns (stream Message) {}
}

message ListReply {
  repeated string all_rooms = 1;
  repeated string joined_rooms = 2;
}

message Request {
  string username = 1;
  repeated string arguments = 2;
}

message Reply {
  string msg = 1;
}

message Message {
  //Username who sent the message
  string username = 1;
  //Message from the user
  string msg = 2;
  //Time the message was sent
  google.protobuf.Timestamp timestamp = 3;
}

message FollowedBy {
  repeated string username = 1;
}